go-controller/pkg/ovn/gateway_init.go

```diff
@@ -425,31 +466,47 @@ func addPolicyBasedRoutes(nodeName, mgmtPortIP string, hostIfAddrs []*net.IPNet)
 	return nil
 }
 
-func (oc *Controller) addNodeLocalNatEntries(node *kapi.Node, mgmtPortMAC, mgmtPortIP string) error {
+func (oc *Controller) addNodeLocalNatEntries(node *kapi.Node, mgmtPortMAC string, mgmtPortIfAddr *net.IPNet) error {
 	var externalIP net.IP
 
+	isIPv6 := utilnet.IsIPv6CIDR(mgmtPortIfAddr)
 	annotationPresent := false
 	externalIPs, err := util.ParseNodeLocalNatIPAnnotation(node)
 	if err == nil {
-		klog.V(5).Infof("Found node local NAT IP %s for the node %s, so reusing it", externalIPs, node.Name)
-		annotationPresent = true
-		externalIP = externalIPs[0]
-	} else {
-		externalIP, err = oc.nodeLocalNatIPAllocator.AllocateNext()
+		for _, ip := range externalIPs {
+			if isIPv6 == utilnet.IsIPv6(ip) {
+				klog.V(5).Infof("Found node local NAT IP %s in %v for the node %s, so reusing it", ip, externalIPs, node.Name)
+				externalIP = ip
+				annotationPresent = true
+				break
+			}
+		}
+	}
+	if !annotationPresent {
+		if isIPv6 {
+			externalIP, err = oc.nodeLocalNatIPv6Allocator.AllocateNext()
+		} else {
+			externalIP, err = oc.nodeLocalNatIPv4Allocator.AllocateNext()
+		}
 		if err != nil {
 			return fmt.Errorf("error allocating node local NAT IP for node %s: %v", node.Name, err)
 		}
+		externalIPs = append(externalIPs, externalIP)
 		defer func() {
 			// Release the allocation on error
 			if err != nil {
-				_ = oc.nodeLocalNatIPAllocator.Release(externalIP)
+				if isIPv6 {
+					_ = oc.nodeLocalNatIPv6Allocator.Release(externalIP)
+				} else {
+					_ = oc.nodeLocalNatIPv4Allocator.Release(externalIP)
+				}
```

Comment:
(Note for later: the ipallocator stuff is awkward and we should make it more natively dual-stack friendly like subnetallocator.)

Reply:
Done - added a note in ovn.go where nodeLocalNatIPv4Allocator & nodeLocalNatIPv6Allocator are defined.